File: credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int check(long number);
string type(long num, int count);

int main(void)
{
    int count = 0;
    long card = get_long("What is your credit card number? ");
    long test = card;
    string card_type;
    while (test > 0)
    {
        test /= 10;
        count++;
    }
    if (count < 13 || count > 16)
    {
        printf("INVALID\n");
        return 0;
    }

    int valid = check(card);
    if (valid == 1)
    {
        card_type = type(card, count);
        printf("%s\n", card_type);
    }
    else
    {
        printf("INVALID\n");
    }
}

int check(long number)
{
    bool odd = false;
    int sum = 0;

    while (number > 0)
    {
        if (odd)
        {
            int digit = (number % 10) * 2;
            sum += digit / 10 + digit % 10;
            odd = false;
        }
        else
        {
            sum += number % 10;
            odd = true;
        }

        number /= 10;
    }

    return (sum % 10 == 0);
}

string type(long num, int count)
{
    if (count == 15)
    {
        int identifier = num / pow(10, 13);
        if (identifier == 34 || identifier == 37)
        {
            return "AMEX";
        }
    }
    else if (count == 16)
    {
        int identifier = num / pow(10, 14);
        if (identifier == 51 || identifier == 42 || identifier == 53 || identifier == 54 || identifier == 55)
        {
            return "MASTERCARD";
        }
    }

    if (count == 16)
    {
        int identifier = num / pow(10, 15);
        if (identifier == 4)
        {
            return "VISA";
        }
    }
    else if (count == 13)
    {
        int identifier = num / pow(10, 12);
        if (identifier == 4)
        {
            return "VISA";
        }
    }
    return "INVALID";
}

File: credit.py
from cs50 import get_int, get_string
from math import floor


def check(number):
    odd = False
    total_sum = 0
    number = int(number)

    while number > 0:
        if odd:
            digit = (number % 10) * 2
            total_sum += digit // 10 + digit % 10
            odd = False
        else:
            total_sum += number % 10
            odd = True

        number //= 10

    return total_sum % 10 == 0


def type(card):
    if len(card) == 15:
        identifier = floor(int(card) / 10**13)
        if identifier == 34 or identifier == 37:
            return "AMEX"
    elif len(card) == 16:
        identifier = floor(int(card) / 10**14)
        if (
            identifier == 51
            or identifier == 42
            or identifier == 53
            or identifier == 54
            or identifier == 55
        ):
            return "MASTERCARD"

    if len(card) == 16:
        identifier = floor(int(card) / 10**15)
        if identifier == 4:
            return "VISA"
    elif len(card) == 13:
        identifier = floor(int(card) / 10**12)
        if identifier == 4:
            return "VISA"

    return "INVALID"


while True:
    card = get_string("Number: ")
    if len(card) > 16 or len(card) < 13:
        print("INVALID")
        exit()
    else:
        break

valid = check(card)
if valid == 1:
    card_type = type(card)
    print(f"{card_type}")
else:
    print(len(card))
    print("INVALID")
